frame/contracts/src/exec.rs

```diff
@@ -582,6 +612,7 @@ where
 		)?;
 		if let Some(ContractInfo::Alive(info)) = ContractInfoOf::<T>::take(&self_id) {
 			Storage::<T>::queue_trie_for_deletion(&info)?;
+			E::remove_user(info.code_hash);
```

Comment:
So if the above `transfer` succeeds but `queue_trie_for_deletion` fails (and thus doesn't `remove_user`), then the whole transaction is rolled back?

Reply:
Great question! Yes it does. Every contract `call` is called under a storage transaction and when it fails all changes are rolled back. When a contract calls `seal_terminate` and it fails the control never returns to the contract and changes re rolled back. Almost all `seal_*` functions behave in this way. Notable exceptions are are `seal_transfer`, `seal_call` and `seal_instantiate` where the caller can decide to carry on when a failure occurs. Extra care must be taken there.

Returning an `Err` from any `seal_*` function in `define_env!` will trigger this trap and rolls back the currents contract changes.

Reply:
Good to know!